fix: Better IPv6 address parsing and fallback

voidr3aper-anon opened the pull request:
Since the project didn't support ipv6 and port perfectly and had some bugs, I have made these changes to fix those. 

Previously if you use ipv6 the value that is being set is null value

and give you error below 

`2025/10/10 13:38:54 Establishing MASQUE connection to <nil>:443`

and also in many situations applications use different ipv6 formats like below that should be supported :

```
2606:4700:103::1
[2606:4700:103::1]
[2606:4700:103::1]:port
```
Also, we saw that defining   a different option to set the port would be really better, and we don't have issues with defining port with ipv6 if we use method 2 of showing the ipv6. 

As I'm a developer of a program called oblivion which you can find in GitHub, and we provide warp GUI for people we found love issues and tried to fix as we can :)

I home the merge request help you, and you merge it, so everyone can use.



The third problem we saw in code, and we tried to fix was the log that TLS handshake fails was incomplete and sometimes with some endpoints happens, and it's because of the invalid token and or one of the keys 

we fix the issue to check that, so the log below won't happen 

`time=2025-10-10T14:20:00.508211361-05:00 level=INFO msg="Failed to connect tunnel: CRYPTO_ERROR 0x128 (remote): tls: handshake failure"`

--In the Name of Freedom

voidr3aper-anon committed:
fix-ipv66-support

Diniboy1123 commented on api/masque.go:
```diff
@@ -59,8 +62,17 @@ func PrepareTlsConfig(privKey *ecdsa.PrivateKey, peerPubKey *ecdsa.PublicKey, ce
 			}
 
 			if !cert.PublicKey.(*ecdsa.PublicKey).Equal(peerPubKey) {
-				// reason is incorrect, but the best I could figure
```
Please keep this comment above the `NoValidChains` note.

Diniboy1123 commented on cmd/socks.go:
```diff
@@ -83,19 +83,38 @@ var socksCmd = &cobra.Command{
-				Port: connectPort,
+			ip, err = config.ParseIPFromEndpoint(config.AppConfig.EndpointV4)
+			if err != nil {
+				// fallback to IPv6 endpoint if IPv4 is missing or unparsable
```
If we do that, we should definitely log the event that we are falling back. Or maybe even introduce a flag to enable fallback manually. I imagine some people wouldn't want to fallback, but connect on a specific endpoint.

Diniboy1123 commented on config/config.go:
```diff
@@ -118,3 +123,65 @@ func (*Config) GetEcEndpointPublicKey() (*ecdsa.PublicKey, error) {
+		}
+	}
+
+	host = strings.Trim(host, "[]")
```
That Trim would just remove all brackets. we can just do something like

```go
func ParseIPFromEndpoint(endpoint string) (net.IP, error) {
	if ip := net.ParseIP(endpoint); ip != nil {
		return ip, nil
	}

	host := endpoint
	if strings.Contains(endpoint, ":") {
		h, _, err := net.SplitHostPort(endpoint)
		if err == nil {
			host = h
		} else if addrErr, ok := err.(*net.AddrError); ok && strings.Contains(addrErr.Err, "missing port") {
		} else {
			host = endpoint
		}
	}

	host = strings.TrimPrefix(host, "[")
	host = strings.TrimSuffix(host, "]")

	if ip := net.ParseIP(host); ip != nil {
		return ip, nil
	}

	return nil, fmt.Errorf("failed to parse IP from endpoint %q: not a valid IPv4/IPv6 literal", endpoint)
}
```

which sounds a little more correct to me regarding what we are trying to do here.

Diniboy1123 reviewed (commented):
Thanks for this. Overall good stuff. Thanks for this! I added a few comments here and there, please address them if you agree.

And also I would appreciate if you would call the commit `fix: Better IPv6 address parsing and fallback` or so.

voidr3aper-anon commented:
i check the code and it was okay overall .i made the changes

Diniboy1123 commented:
I see that you resolved the conversation points, but probably forgot to push the changes? Even the commit message is the old one. So please address those and then it's going to be ready for merge.

voidr3aper-anon committed:
Trim brackets from host string in config

voidr3aper-anon committed:
fix: Better IPv6 address parsing and fallback

voidr3aper-anon commented:
> I see that you resolved the conversation points, but probably forgot to push the changes? Even the commit message is the old one. So please address those and then it's going to be ready for merge.

When   you click on resolve thread it pushes the changes in same commit, and you can see them in changes not by looking in commit files. 

I agree to trim and added the changes you provided. It's ready to go. (for complete changes, check the changes tab)
I added new commit for the commit massage though :)

Diniboy1123 commented on config/config.go:
```diff
@@ -118,3 +123,66 @@ func (*Config) GetEcEndpointPublicKey() (*ecdsa.PublicKey, error) {
+			host = h
+		} else {
+			// If SplitHostPort failed, try trimming brackets (in case it's "[ip]" or similar)
+			host = strings.Trim(endpoint, "[]")
```
This logic still seems to trim all occurences of brackets? Or am I missing something?

Diniboy1123 commented on api/masque.go:
```diff
@@ -59,8 +62,17 @@ func PrepareTlsConfig(privKey *ecdsa.PrivateKey, peerPubKey *ecdsa.PublicKey, ce
 			}
 
 			if !cert.PublicKey.(*ecdsa.PublicKey).Equal(peerPubKey) {
-				// reason is incorrect, but the best I could figure
```
This comment still shows up to me as gone.

voidr3aper-anon replied:
I think now it's okay. I don't know how, but some changes hadn't been pushed successfully !
If any issue persist, please inform me :)

Diniboy1123 reviewed (commented):
I still see some weird stuff hence why my confusion. Once those are addressed to, I will really merge. :)

Thanks

voidr3aper-anon committed:
fix: Better IPv6 address parsing and fallback

voidr3aper-anon commented:
Check it again, please. I think the issue is solved :)

Adm0 commented:
Why not use "net/netip" parsing functions?
And don't add fields to the config.

If someone needs to specify a port in the config,  it can be done like this:
"endpoint_v4": "162.159.198.1:500",
"endpoint_v6": "[2606:4700:103::1]:500",

internal/utils.go:
```go
import 	"net/netip"

func ParseAddrPort(endpoint string, defaultPort uint16) (netip.AddrPort, error) {
	addr, err := netip.ParseAddr(endpoint)
	if err != nil {
		return netip.ParseAddrPort(endpoint)
	}
	return netip.AddrPortFrom(addr, defaultPort), nil
}
```
cmd:
```go
import 	"net/netip"

...
		var addr netip.AddrPort
		if ipv6, err := cmd.Flags().GetBool("ipv6"); err == nil && !ipv6 {
			addr, err = internal.ParseAddrPort(config.AppConfig.EndpointV4, 443)
			if err != nil {
				cmd.Printf("Failed to parse IPv4 endpoint: %v", err)
				addr, err = internal.ParseAddrPort(config.AppConfig.EndpointV6, 443)
				if err != nil {
					cmd.Printf("Failed to parse IPv6 endpoint: %v", err)
					return
				}
			}
		} else {
			addr, err = internal.ParseAddrPort(config.AppConfig.EndpointV6, 443)
			if err != nil {
				cmd.Printf("Failed to parse IPv6 endpoint: %v", err)
				addr, err = internal.ParseAddrPort(config.AppConfig.EndpointV4, 443)
				if err != nil {
					cmd.Printf("Failed to parse IPv4 endpoint: %v", err)
					return
				}
			}
		}
		endpoint := net.UDPAddrFromAddrPort(addr)

		if cmd.Flags().Changed("connect-port") {
			endpoint.Port, err = cmd.Flags().GetInt("connect-port")
			if err != nil {
				cmd.Printf("Failed to get connect port: %v\n", err)
				return
			}
		}
...
```

voidr3aper-anon commented:
The previous config I had does the same thing. But because the app uses IPv4 port, I did the same mechanism. Also, people write scanners to dynamically change the port, but the single IP stays, and it's not bad actually.
